crates/pop-contracts/src/up.rs

```diff
@@ -93,9 +102,95 @@ pub async fn set_up_upload(
 
 	let upload_exec: UploadExec<DefaultConfig, DefaultEnvironment, Keypair> =
 		UploadCommandBuilder::new(extrinsic_opts).done().await?;
+
 	Ok(upload_exec)
 }
 
+pub async fn get_upload_payload(up_opts: UpOpts) -> anyhow::Result<Vec<u8>> {
+	let code = get_contract_code(up_opts.path.as_ref()).await?;
+
+	let storage_deposit_limit: Option<u128> = None;
+	let upload_code = contract_extrinsics::extrinsic_calls::UploadCode::new(
+		code,
+		storage_deposit_limit,
+		contract_extrinsics::upload::Determinism::Enforced,
+	);
+
+	// TODO: review placement
+	let rpc_client = subxt::backend::rpc::RpcClient::from_url(up_opts.url.as_str()).await?;
+	let client = subxt::OnlineClient::<SubstrateConfig>::from_rpc_client(rpc_client).await?;
+
+	let call_data = upload_code.build();
+	let mut encoded_data = Vec::<u8>::new();
+	call_data.encode_call_data_to(&client.metadata(), &mut encoded_data)?;
+	Ok(encoded_data)
+}
+
+pub async fn get_instantiate_payload(
+	instantiate_exec: InstantiateExec<DefaultConfig, DefaultEnvironment, Keypair>,
+	gas_limit: Weight,
+) -> anyhow::Result<Vec<u8>> {
+	let storage_deposit_limit: Option<u128> = None;
+	let mut encoded_data = Vec::<u8>::new();
+	match instantiate_exec.args().code() {
+		contract_extrinsics::Code::Upload(code) => InstantiateWithCode::new(
+			instantiate_exec.args().value(),
+			gas_limit,
+			storage_deposit_limit,
+			code.clone(),
+			instantiate_exec.args().data().into(),
+			instantiate_exec.args().salt().into(),
+		)
+		.build()
+		.encode_call_data_to(&instantiate_exec.client().metadata(), &mut encoded_data),
+		contract_extrinsics::Code::Existing(hash) => Instantiate::new(
+			instantiate_exec.args().value(),
+			gas_limit,
+			storage_deposit_limit,
+			hash.clone(),
+			instantiate_exec.args().data().into(),
+			instantiate_exec.args().salt().into(),
+		)
+		.build()
+		.encode_call_data_to(&instantiate_exec.client().metadata(), &mut encoded_data),
+	}?;
+
+	Ok(encoded_data)
+}
+
+pub async fn get_contract_code(
+	path: Option<&PathBuf>,
+) -> anyhow::Result<contract_extrinsics::WasmCode> {
+	let manifest_path = get_manifest_path(path.map(|p| p as &Path))?;
+
+	// signer does not matter for this
+	let signer = create_signer("//Alice")?;
+	let extrinsic_opts =
+		ExtrinsicOptsBuilder::<DefaultConfig, DefaultEnvironment, Keypair>::new(signer)
+			.manifest_path(Some(manifest_path))
+			.done();
+	let artifacts = extrinsic_opts.contract_artifacts()?;
+
+	let artifacts_path = artifacts.artifact_path().to_path_buf();
+	let code = artifacts.code.ok_or_else(|| {
+		anyhow::anyhow!("Contract code not found from artifact file {}", artifacts_path.display())
```

Comment:
In the crates return a custom Error instead of `anyhow`

Reply:
thanks for the pointer!